### PR TITLE
Create extra branch

### PR DESCRIPTION
**Pull request by ruslankotliar**

### CODE REVIEW

File: packages/backend/src/config/database.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable no-console */
+import { ConnectionOptions, connect } from 'mongoose';
+
+const connectDB = async () => {
+  try {
+    const options: ConnectionOptions = {
+      useNewUrlParser: true,
```
Comment (SwiLink) [inline packages/backend/src/config/database.ts:7]: Try to create here your local database, connect to it, interact with and create keys for connection

File: packages/backend/src/controllers/todo.controller.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  UpdateTodoBody,
+  DeleteTodoById,
+  UpdateTodoParams
+} from '../types/todos.type';
+
+export class TodoController {
+  constructor(private todoService: TodoService) {}
+
+  async getAllTodo(req: TypedRequestParams<GetById>) {
+    const { id } = req.params;
```
Comment (SwiLink) [inline packages/backend/src/controllers/todo.controller.ts]: Not necessary, but recommended not to pull all request, just destructure req from start and better to type return type of function/method.

File: packages/backend/src/controllers/todo.controller.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  async createOneTodo(req: TypedRequestBody<CreateTodo>) {
+    const { body } = req;
+    const newTodo = await this.todoService.createOne(body);
+    return newTodo;
+  }
+
+  async updateOneTodo(
+    req: TypedRequestParams<UpdateTodoParams> & TypedRequestBody<UpdateTodoBody>
```
Comment (SwiLink) [inline packages/backend/src/controllers/todo.controller.ts]: Weird type, i think it needs to be one type for all req, read about `Request` type in `express`

File: packages/backend/src/helpers/query.helper.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import { Filters, Pagination } from '../types/todos.type';
+
+const setFilters = (access: string, status: string, query: string): Filters => {
+  let filters: Filters = { private: false };
+  if (status === 'all') delete filters.completed;
```
Comment (SwiLink) [inline packages/backend/src/helpers/query.helper.ts]: Better to use `switch-case` here, if it's impossible with dynamic data

File: packages/backend/src/middlewares/validation/auth.middleware.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    // check whether user is unique
+    const user = await User.findOne({ email });
+    if (user) throw new Error('Conflict');
+
+    // hash password
+    const hashedPassword = await hashPassword(password);
+
+    // send data in controller
+    req.body.email = email;
```
Comment (SwiLink) [inline packages/backend/src/middlewares/validation/auth.middleware.ts]: Why you need all this data in each method of each controller? It's dangereously

File: packages/backend/src/routes/api/todos.route.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+todosRouter.get('/get-todo/:id', sendResponse(todoController.getOneTodo.bind(todoController)));
+todosRouter.post(
+  '/create-todo',
+  validateBody<ITodo>(taskSchema),
+  sendResponse(todoController.createOneTodo.bind(todoController))
+);
+todosRouter.put(
+  '/update-todo/:id',
+  checkExists<ITodo>(Todo),
+  validateBody<ITodo>(taskSchema),
```
Comment (SwiLink) [inline packages/backend/src/routes/api/todos.route.ts:19]: Understood, for what it's created. Read about mappers on backend, not bad implementation too.

File: packages/backend/src/server.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import express from 'express';
+import 'dotenv/config';
+import AppRouter from './routes';
+import connectDB from './config/database';
+import { errorHandler } from './middlewares/index';
+
+const app = express();
+const router = new AppRouter(app);
+// Connect to MongoDB
+connectDB();
```
Comment (SwiLink) [inline packages/backend/src/server.ts:12]: Doesn't need to be here

File: packages/backend/src/types/todos.type.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+};
+
+type Filters = {
+  private?: Boolean;
+  completed?: Boolean;
+  $text?: { $search: string; $caseSensitive: Boolean; $diacriticSensitive: Boolean };
+};
+
+type Pagination = [
+  { $setWindowFields: { output: { totalCount: { $count: {} } } } },
```
Comment (SwiLink) [inline packages/backend/src/types/todos.type.ts]: It's not the deal of frontend to construct such structures, only at backend after resolver in your situation

File: packages/frontend/public/manifest.json [hunk 9]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (SwiLink) [inline packages/frontend/public/manifest.json:1]: If don't know, read about, why you need this file.

File: packages/frontend/src/modules/common/components/table/table-desktop/table-desktop.component.tsx [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  CustomViewButton
+} from '../table-todos.styled';
+
+export const TodosDesktopComponent = ({
+  todos = [],
+  updateTodoMutation,
+  removeTodoMutation
+}: any) => (
+  <CustomPaper>
+    <TableContainer>
```
Comment (SwiLink) [inline packages/frontend/src/modules/common/components/table/table-desktop/table-desktop.component.tsx]: Weird and a lot of code, but ok, it's another library. Try `antd`, for example and look how better to configure table

File: packages/frontend/src/modules/common/components/table/table-layout/table-layout.component.tsx [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+  CustomContainer,
+  CustomForm,
+  CustomFormControl,
+  CustomTab
+} from './table-layout.styled';
+import { useQuery } from '../../../hooks';
+import { debounce } from '../../../helpers';
+
+export const TodosTableLayout = ({ children, setTrigger, count }: any) => {
+  const [searchParams, setSearchParams] = useSearchParams();
```
Comment (SwiLink) [inline packages/frontend/src/modules/common/components/table/table-layout/table-layout.component.tsx]: Bad implementation, you do all this static, it needs to be dynamic, you need to create helper for right preparing data for url params independent from its variables, and then pass it to url and do a request for data.

File: packages/frontend/src/modules/common/components/table/table-todos.component.tsx [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    isError: isErrorMutate,
+    error: errorMutate,
+    allTodos,
+    setTrigger,
+    updateTodoMutation,
+    removeTodoMutation
+  } = useTodos();
+  const { isLoading, isError, error, data: todos } = allTodos;
+
+  if (isLoading) return <SpinnerComponent />;
```
Comment (SwiLink) [inline packages/frontend/src/modules/common/components/table/table-todos.component.tsx]: You can do it inside JSX with ternary operator `... ? (...) : ... ? (...) : (...)`

File: packages/frontend/src/modules/common/components/todo-form/todo-form.component.tsx [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    createTodoMutation
+  } = useTodos();
+  const { isLoading, isError, error, data: todo } = singleTodo;
+
+  if (isLoading) return <SpinnerComponent />;
+  if (isError || isErrorMutate) return <ErrorComponent error={error || errorMutate} />;
+  return (
+    <Box>
+      <Typography variant="h3">{todo ? 'Update todo' : 'Create todo'}</Typography>
+      <Formik
```
Comment (SwiLink) [inline packages/frontend/src/modules/common/components/todo-form/todo-form.component.tsx]: It's to your wish, but i prefer to initialize `Formik` outside JSX.

File: packages/frontend/src/modules/common/components/todo-form/todo-form.component.tsx [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        }}
+        validationSchema={todoSchema}
+        onSubmit={(values, { setSubmitting }) => {
+          todo ? updateTodoMutation(values) : createTodoMutation(values);
+          setSubmitting(false);
+        }}
+      >
+        {({ submitForm, isSubmitting }) => (
+          <Form>
+            <Field
```
Comment (SwiLink) [inline packages/frontend/src/modules/common/components/todo-form/todo-form.component.tsx]: And it's not very necessary to use this `Field` and `Form` from `Formik`, it just heaps the component.

File: packages/frontend/src/modules/common/consts/app-keys.const.ts [hunk 15]
```diff
@@ -0,0 +1,45 @@
+// Local storage keys
```
Comment (SwiLink) [inline packages/frontend/src/modules/common/consts/app-keys.const.ts:1]: Good thing, when you are writing code, always think about avoiding magic numbers and magic phrases, move it to constants

File: packages/frontend/src/modules/common/schemas/todo.schema.ts [hunk 16]
```diff
@@ -0,0 +1,9 @@
+import * as Yup from 'yup';
```
Comment (SwiLink) [inline packages/frontend/src/modules/common/schemas/todo.schema.ts:1]: I don't think that you need to move it so far away from component, you can declare it just outside component after imports if it doesn't need any data from component for conditions

File: packages/frontend/src/modules/navigation/index.tsx [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+import UpdateUserPageComponent from '../user-update';
+import TodoFormPageComponent from '../todo-form';
+
+import { APP_KEYS } from '../common/consts';
+import { ProtectedRoute } from '../common/components/route-protected';
+
+export const MainRouter = () => (
+  <Router>
+    <Routes>
+      <Route element={<ProtectedRoute />}>
```
Comment (SwiLink) [inline packages/frontend/src/modules/navigation/index.tsx]: If you are creating modules, better to create routes file in each file and then concatenate in one main, then import it here and map through

File: packages/backend/src/server.ts (hunk 7 above)
Comment (SwiLink) [reply packages/backend/src/server.ts:12]: Better to connect in each model just to separate collection.

**Review by SwiLink — COMMENTED** [review]
SwiLink

File: packages/backend/src/config/database.ts (hunk 1 above)
Comment (ruslankotliar) [reply packages/backend/src/config/database.ts:7]: I will do it somewhen in another project

File: packages/backend/src/controllers/todo.controller.ts (hunk 2 above)
Comment (ruslankotliar) [reply packages/backend/src/controllers/todo.controller.ts]: Done

File: packages/backend/src/controllers/todo.controller.ts (hunk 3 above)
Comment (ruslankotliar) [reply packages/backend/src/controllers/todo.controller.ts]: Nope, that is alright
Here is a great source: https://plainenglish.io/blog/typed-express-request-and-response-with-typescript

File: packages/backend/src/helpers/query.helper.ts (hunk 4 above)
Comment (ruslankotliar) [reply packages/backend/src/helpers/query.helper.ts]: Done

**Pull request merged**